### PR TITLE
Switch to SQLModel Database stuff 

### PR DESCRIPTION
**Pull request by jontyms**
TODO:

- [x] rework config options 
- [x] Alembic (db migrations
- [x] Testing
- [x] Remove Print statements
- [x] Code review
- [x] Even more Testing

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -22,7 +22,7 @@ RUN mv bws /usr/local/bin
 RUN rm -r /tmp/
 
 # Install the dependencies
-RUN pip install --no-cache-dir -r requirements.txt
+RUN --mount=type=cache,target=/root/.cache/pip pip install -r requirements.txt
```
Comment (Helithumper) [inline Dockerfile:25]: Unsure if this is a good idea or not but haven't seen it used before

File: Dockerfile-testing [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+# Copy the requirements file to the container
+COPY requirements.txt .
+
+# Install build-essential
+RUN apt-get update && apt-get install -y build-essential
+
+# Clean up
+RUN apt-get clean && rm -rf /var/lib/apt/lists/*
+
+ADD https://github.com/bitwarden/sdk/releases/download/bws-v0.4.0/bws-x86_64-unknown-linux-gnu-0.4.0.zip /tmp
```
Comment (Helithumper) [inline Dockerfile-testing:16]: Is this the official installation method for the bitwarden SDK?

File: app/main.py [hunk 3]
```diff
@@ -194,36 +198,29 @@ async def oauth_transformer_new(
...
 
     # Generate a new user ID or reuse an existing one.
-    query_for_id = table.scan(
-        FilterExpression=Attr("discord_id").eq(str(discordData["id"]))
-    )
-    query_for_id = query_for_id.get("Items")
-
+    statement = select(UserModel).where(UserModel.discord_id == discordData["id"])
+    user = session.exec(statement).one_or_none()
+    print(user)
```
Comment (Helithumper) [inline app/main.py:203]: Replace with logger usage

File: app/main.py [hunk 4]
```diff
@@ -243,54 +240,26 @@ async def oauth_transformer_new(
...
-    # Push data back to DynamoDB
-    if is_new:
-        table.put_item(Item=full_data)
-    else:
-        table.update_item(
-            Key={"id": member_id},
-            UpdateExpression="SET discord = :discord",
-            ExpressionAttributeValues={":discord": full_data["discord"]},
-        )
+            "color": discordData.get("accent_color"),
```
Comment (Helithumper) [inline app/main.py:249]: Should these default to something other than `None`?

File: app/main.py [hunk 5]
```diff
@@ -368,11 +337,16 @@ async def forms(
...
         )
 
-
-    # Get data from DynamoDB
-    user_data = table.get_item(Key={"id": user_jwt.get("id")}).get("Item", None)
-
+    # Get data from SqlModel
+    statement = (
+        select(UserModel)
+        .where(UserModel.id == user_jwt["id"])
```
Comment (Helithumper) [inline app/main.py:343]: Handling of `KeyError`?

File: app/models/user.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
+import re
+import uuid
+from typing import Optional
+
+from pydantic import BaseModel, validator
+from sqlmodel import Field, Relationship, SQLModel
+
+
+class DiscordModel(SQLModel, table=True):
+    id: Optional[int] = Field(default=None, primary_key=True)
```
Comment (Helithumper) [inline app/models/user.py:10]: Is it a good idea to make all these fields nullable? Why not just make the reference to this table nullable instead?

File: app/models/user.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+    pirate: Optional[bool] = False
+    host_at_ucf: Optional[bool] = False
+    signtime: Optional[int] = 0
+
+    user_id: Optional[uuid.UUID] = Field(default=None, foreign_key="usermodel.id")
+    user: "UserModel" = Relationship(back_populates="ethics_form")
+
+
+
+class UserModel(SQLModel, table=True):
```
Comment (Helithumper) [inline app/models/user.py:40]: Should almost all of these fields be nullable?

File: app/routes/api.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+from app.util.errors import Errors
+from app.util.forms import Forms, apply_fuzzy_parsing
+from app.util.kennelish import Transformer
+
+logger = logging.getLogger(__name__)
+
+router = APIRouter(prefix="/api", tags=["API"], responses=Errors.basic_http())
+
+
+"""
```
Comment (Helithumper) [inline app/routes/api.py:23]: Docstrings go under the function signature

File: app/routes/api.py [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+    statement = select(UserModel).where(UserModel.id == user_id)
+    result = session.exec(statement)
+    user = result.one_or_none()
+
+    if not user:
+        raise HTTPException(status_code=404, detail="User not found")
+
+    # Update the ethics form with new values
+    validated_data = apply_fuzzy_parsing(ethics_form_data.model_dump(exclude_unset=True), EthicsFormModel)
+    print(validated_data.dict())
```
Comment (Helithumper) [inline app/routes/api.py:114]: Logger

File: app/util/database.py [hunk 10]
```diff
@@ -0,0 +1,17 @@
+
+# Create the database
+from sqlmodel import Session, SQLModel, create_engine
+from sqlmodel.pool import StaticPool
+
+DATABASE_URL = "sqlite:///:memory:"
```
Comment (Helithumper) [inline app/util/database.py:6]: Read from ENVVAR w/ Defaults

**Comment by Helithumper** [comment]
We'll need some sort of migration system for at least up-migrations on the SQL DB

**Comment by Helithumper** [comment]
Did a quick sweep with some notes, Overall a good first step but needs some polish 🥳

File: Dockerfile [hunk 11]
```diff
@@ -22,7 +22,7 @@ RUN mv bws /usr/local/bin
...
 
 # Copy the application code to the container
 COPY . .
```
Comment (Helithumper) [inline Dockerfile:28]: Include a .Dockerignore for files that don't need to be in the container

File: Dockerfile-testing [hunk 12]
```diff
@@ -0,0 +1,34 @@
+# Use the official Python base image
+FROM python:3.11-bookworm
+
+# Set the working directory in the container
+WORKDIR /app
+
+# Copy the requirements file to the container
+COPY requirements.txt .
```
Comment (Helithumper) [inline Dockerfile-testing:8]: Move this further down, any changes to requirements will cause the apt execution to restart

File: docker-compose.yml [hunk 13]
```diff
@@ -8,6 +8,7 @@ services:
       - 8000:8000
     volumes:
       - ./config/options.yml:/app/config/options.yml
+      - ./database/:/app/database
```
Comment (Helithumper) [inline docker-compose.yml:11]: Nit: usually this is called `./data`

File: Dockerfile (hunk 1 above)
Comment (jontyms) [reply Dockerfile:25]: This create a pip cache that is stored outside of the docker container, speeding up rebuilds 
https://pythonspeed.com/articles/docker-cache-pip-downloads/

File: app/main.py (hunk 5 above)
Comment (jontyms) [reply app/main.py:343]: Should not occur @Authentication.member will fail if the user has an invalid JWT

File: Dockerfile (hunk 11 above)
Comment (jontyms) [reply Dockerfile:28]: Speeds up rebuilds that require a pip install by caching the dependencies  

https://pythonspeed.com/articles/docker-cache-pip-downloads/

File: app/models/user.py (hunk 7 above)
Comment (jontyms) [reply app/models/user.py:39]: I will review, but most of the values are unknown when we first write the user to the DB before they do the onboarding forms

File: app/models/user.py (hunk 6 above)
Comment (jontyms) [reply app/models/user.py:10]: Probably not

File: Dockerfile-testing (hunk 2 above)
Comment (jontyms) [reply Dockerfile-testing:16]: Yes https://bitwarden.com/help/secrets-manager-cli/

I didn't see they had an sdk before might switch to that https://bitwarden.com/help/secrets-manager-sdk/

File: Dockerfile (hunk 1 above)
Comment (jontyms) [reply Dockerfile:25]: Speeds up rebuilds that require a pip install by caching the dependencies

https://pythonspeed.com/articles/docker-cache-pip-downloads/

**Comment by jontyms** [comment]
> We'll need some sort of migration system for at least up-migrations on the SQL DB

https://testdriven.io/blog/fastapi-sqlmodel/

most likely we will use Alembic

File: docker-compose.yml (hunk 13 above)
Comment (jontyms) [reply docker-compose.yml:11]: #95

File: Dockerfile-testing (hunk 12 above)
Comment (jontyms) [reply Dockerfile-testing:8]: #95

File: Dockerfile (hunk 11 above)
Comment (jontyms) [reply Dockerfile:28]: #95

File: Dockerfile-testing (hunk 2 above)
Comment (jontyms) [reply Dockerfile-testing:16]: #96 #95

**Comment by Helithumper** [comment]
Alembic's fine, that works

File: Dockerfile (hunk 1 above)
Comment (zhiyan114) [reply Dockerfile:25]: Just looking at random repo and came across this. If you want a faster build, you can always setup CI pipeline with GH Action and store it on PKG registry, then fetch it on the server.

**Pull request merged**